sw/device/silicon_creator/lib/drivers/otbn.c

```diff
@@ -81,8 +81,44 @@ static void otbn_write(uint32_t dest_addr, const uint32_t *src,
   HARDENED_CHECK_EQ(iter_cnt, num_words);
 }
 
-void otbn_execute(void) {
-  abs_mmio_write32(kBase + OTBN_CMD_REG_OFFSET, kOtbnCmdExecute);
+/**
+ * Helper function for running an OTBN command.
+ *
+ * This function blocks until OTBN is idle.
+ *
+ * @param cmd OTBN command.
+ * @param exp_status Expected OTBN status after issuing the command.
+ * @param error Error to return if operation fails.
+ * @return Result of the operation.
+ */
+static rom_error_t otbn_cmd_run(otbn_cmd_t cmd, otbn_status_t exp_status,
+                                rom_error_t error) {
+  abs_mmio_write32(kBase + OTBN_CMD_REG_OFFSET, cmd);
+
+  // Read back status register to confirm that the command is being run.
+  rom_error_t res = kErrorOk ^ exp_status;
+  res ^= abs_mmio_read32(kBase + OTBN_STATUS_REG_OFFSET);
+  if (launder32(res) != kErrorOk) {
+    return error;
+  }
+  HARDENED_CHECK_EQ(res, kErrorOk);
```

Comment:
This is a bit awkward, because it introduces a race if the command happens to run really fast. Practically speaking, I doubt it will matter (because our secure wipe takes ~100 cycles), but it doesn't quite seem like the right thing to do. Is there a specific situation that it's trying to catch?

Reply:
You're right, this could result in a race condition if the commands ran really fast but I was also thinking that this wouldn't matter in practice.

According to the HWIP spec, invalid commands are ignored and CMD register is write-only. This piece of code tries to confirm that we were able to issue the command that we intended. PLMK if you think we should be doing something else.

Reply:
Ah, I see. Another thing you could do is completely skip this check and move right on to the "is OTBN idle yet?" check that happens below. If we'd written some invalid command then we'd never leave idle. I assume that whatever calls this to run an OTBN operation will be checking some "did I do it?" flag upon completion?

But that's not really going to work for the memory key rotation, of course. I guess that if we can be completely certain we'll be quick enough on the CPU (and won't have to service interrupts etc...) then maybe this patch already has the right behaviour?

Obviously, none of these approaches (including what we've got here) work if there's a possibility that OTBN wasn't idle before we started. I guess we're ensuring that by making sure that silicon_creator only communicates with OTBN through these driver routines?

Reply:
> I guess that if we can be completely certain we'll be quick enough on the CPU (and won't have to service interrupts etc...) then maybe this patch already has the right behaviour?
>
> Obviously, none of these approaches (including what we've got here) work if there's a possibility that OTBN wasn't idle before we started. I guess we're ensuring that by making sure that silicon_creator only communicates with OTBN through these driver routines?

Yes, that's what I'm thinking, since this code is exclusively for mask rom and rom_ext and we won't have anything else going on. I'll add a comment for this assumption.